### PR TITLE
Add tmp_to_dst_file context manager to Dataset class

### LINKED ISSUE #148: Use local scr for initial downloads / heavy writes
As a best practice given the potential scale of many of our pipelines (i.e., potentially running dozens+ tasks across nodes), I'd like to minimize the continuous IO on our main file system. This should be relatively simple by just using the node's local disk for downloads / processing before copying to the main file system. Even in scenarios where operations are quick and not heavy IO, this shouldn't slow jobs down much and worth the extra piece of mind.

(Currently, heavy IO is seemingly causing extra issues on our aging file system, but we are in the process of moving everything to a brand new file system.)

### PR DESCRIPTION
**Pull request by jacobwhall**
`tmp_to_dst_file()` takes final_dst argument, yields a temporary file path that you can write to. When the context manager exits, it moves the temporary file to final_dst

Resolves #148

### CODE REVIEW

File: global_scripts/dataset.py [hunk 1]
```diff
@@ -9,6 +10,26 @@
...
+        tmp_dir = Path("/local/scr") / os.environ["USER"] / "TMPDIR"
+        if tmp_dir.exists():
+            return tmp_dir.as_posix()
+    except:
+        pass
+
+    try:
+        return _get_default_tempdir()
+    except FileNotFoundError:
+        raise FileNotFoundError("Unable to find a suitable temporary directory. Please specify tmp_dir when calling run")
```
Comment (jacobwhall) [inline global_scripts/dataset.py:32]: @sgoodm how do you feel about this? I like adding /local/src/$USER/TMPDIR as the default, but having a whole function to find a temporary directory feels unwieldy. I'm also hesitant about my usag e of the internal function `_get_default_tempdir()` from `tempfile`.

**Comment by jacobwhall** [comment]
I've tested the basic functionality of this PR with cru_ts. [Here](https://github.com/jacobwhall/geo-datasets/blob/b5844dc69ddca4de662f842e60bf999f4e5b5944/cru_ts/main.py#L112-L123) is an example of it being used.

File: global_scripts/dataset.py (hunk 1 above)
Comment (jacobwhall) [reply global_scripts/dataset.py:32]: I like your approach here. I'll remove find_tmp_dir() in favor of $TMPDIR, and add the dir kwarg to tmp_to_dst_file.

When first I wrote this PR I assumed that when using the "hpc" task runner with Prefect, worker processes would default to the same $TMPDIR set on the node running the agent. I agree it's worth testing, but the more I think about it the less likely I think the environment variable would be transferred.

**Comment by sgoodm** [comment]
Slight issue was encountered when using an NFS mounted dir for the tmp directory where the .nfs file cannot be removed from the tmpdir and causes the tempdir context manager's dir cleanup to fail and raise an exception (rmtree fails on .nfs file w/ `OSError: [Errno 16] Device or resource busy: '.nfs000` error). See MRC below w/ originall CM commented out and solution implemented.

99% solution seems to be just manually creating the tmpdir ourselves (rather than using a context manager). Since the tmp_to_dst_file context manager is only for a single file, which we move from the tmp_dir, there isn't really a need to delete the full tmp dir.

```
import os 
import time
import shutil
from tempfile import TemporaryDirectory, mkstemp, mkdtemp
from contextlib import contextmanager

@contextmanager
def tmp_to_dst_file(final_dst, tmp_dir=None):
    # with TemporaryDirectory(dir=tmp_dir) as tmp_sub_dir:
    tmp_sub_dir = mkdtemp(dir=tmp_dir)
    tmp_file, tmp_path = mkstemp(dir=tmp_sub_dir)
    # tmp_file.close()
    print(tmp_file, tmp_path)
    print(f"Created temporary file {tmp_path} with final destination {str(final_dst)}")
    yield tmp_path
    try:
        shutil.move(tmp_path, final_dst)
        # shutil.remove(tmp_path)
    except:
        print(f"Failed to transfer temporary file {tmp_path} to final destination {str(final_dst)}")
    else:
        print(f"Successfully transferred {tmp_path} to final destination {str(final_dst)}")
        # tmp_file.close()
        # shutil.rmtree(tmp_sub_dir)
        time.sleep(5)
        print(os.listdir(tmp_sub_dir))
        print(tmp_sub_dir)


final_dl_path = '/sciclone/home20/smgoodman/test.txt'
# tmp_dir = '/sciclone/home20/smgoodman/'
tmp_dir = None


with tmp_to_dst_file(final_dl_path, tmp_dir=tmp_dir) as dl_dst:
    print(f"Starting")
    try:
        print(f"Trying")    
        with open(dl_dst, 'w') as f:
            f.write('test')
    except:
        print(f"Failed")
    else:
        print(f"Success")


# when using scr20, tmpdir still has .nfs file in it until python session ends
# print(os.listdir(tmp_sub_dir))
```

**Pull request merged**